[DropUnnecessaryAssumes] Make the ephemeral value check more precise

nikic opened the pull request:
The initial implementation used a very crude check where a value was considered ephemeral if it has only one use. This is insufficient if there are multiple assumes acting on the same value, or in more complex cases like cyclic phis.

Generalize this to a more typical ephemeral value check, i.e. make sure that all transitive users are in assumes, while stopping at side-effecting instructions.

Interestingly, this doesn't seem to actually matter in practice (no difference on llvm-opt-benchmark).

nikic committed:
[DropUnnecessaryAssumes] Make the ephemeral value check more precise

The initial implementation used a very crude check where a value
was considered ephemeral if it has only one use. This is insufficient
if there are multiple assumes acting on the same value, or in more
complex cases like cyclic phis.

Generalize this to a more typical ephemeral value check, i.e. make
sure that all transitive users are in assumes, while stopping at
side-effecting instructions.

artagnon reviewed (commented):
> Interestingly, this doesn't seem to actually matter in practice (no difference on llvm-opt-benchmark).

Not sure how I feel about this. I've personally abandoned patches in the past due to zero-diff (or even small diff) on llvm-opt-benchmark. The patch is certainly nice for completeness though?

nikic commented:
> > Interestingly, this doesn't seem to actually matter in practice (no difference on llvm-opt-benchmark).
> 
> Not sure how I feel about this. I've personally abandoned patches in the past due to zero-diff (or even small diff) on llvm-opt-benchmark. The patch is certainly nice for completeness though?

Generally agreed. I do think it's valuable to have this to make sure the pass behavior is consistent (i.e. the pass shouldn't break just because we added a second dead assume).

artagnon commented on llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp:
```diff
@@ -17,13 +18,44 @@ using namespace llvm;
+        return false;
+
+  for (unsigned Idx = 0; Idx < Worklist.size(); ++Idx) {
+    auto *I = dyn_cast<Instruction>(Worklist[Idx]);
+    if (!I)
+      return false;
```
Might be preferable to have an Instruction-worklist, and dyn_cast when adding to the Worklist? Also, the AddUser is a bit awkward: just a thought, but maybe it would be valuable to have something like collectUsersRecusively (from VPlan), so we can bail out at collection time? I think the function could also be used in several other places like the new HashRecognize? Then, we can simply check that all_of them are either assumes, or don't have side-effects?

```cpp
static SmallVector<VPUser *> collectUsersRecursively(VPValue *V) {
  SetVector<VPUser *> Users(llvm::from_range, V->users());
  for (unsigned I = 0; I != Users.size(); ++I) {
    VPRecipeBase *Cur = cast<VPRecipeBase>(Users[I]);
    if (isa<VPHeaderPHIRecipe>(Cur))
      continue;
    for (VPValue *V : Cur->definedValues())
      Users.insert_range(V->users());
  }
  return Users.takeVector();
}
```

nikic replied:
I'm generally not a fan of fully flattening users up-front, as it means we can't have early exits.

artagnon commented on llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp:
```diff
@@ -17,13 +18,44 @@ using namespace llvm;
+      continue;
+
+    // Use in side-effecting instruction is non-ephemeral.
+    if (I->mayHaveSideEffects() || I->isTerminator())
```
Curiously, the normal pattern is to check for phis when collecting users, to avoid cycles: wondering why you choose to check terminators?

nikic replied:
This follows our standard definition of what an ephemeral value is, see for example: https://github.com/llvm/llvm-project/blob/1e302e942cdbb2db902530144d0c3891f811168b/llvm/lib/Analysis/CodeMetrics.cpp#L38

I'm specifically not checking for phis here to also allow ephemeral cycles.

artagnon commented on llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp:
```diff
@@ -17,13 +18,44 @@ using namespace llvm;
+    return true;
+  };
+
+  for (Value *V : Affected)
+    for (User *U : V->users())
+      if (!AddUser(U))
+        return false;
```
Hm, what if the Affected value isn't an Instruction? Wouldn't it be better to append the affected values to the Worklist, and let the code that follows dig the users?

nikic replied:
That would be simpler, but it wouldn't do the same thing. In particular, if the affected value itself is side-effecting, then we still want to treat it as ephemeral if it has no other uses. If we just pushed everything to the worklist, we'd treat them as non-ephemeral.

artagnon replied:
Maybe add a quick comment saying that all affected values should be treated ephemeral?

artagnon reviewed (commented):
Okay, let's use the opportunity to think about some ideas then: could you kindly think about collectUsersRecursively?

andjo403 commented on llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp:
```diff
@@ -17,13 +18,44 @@ using namespace llvm;
+  // by assumes. In that case, the assume does not provide useful information.
+  // Note that additional users may appear as a result of inlining and CSE,
+  // so we should only make this assumption late in the optimization pipeline.
+  SmallSetVector<User *, 16> Worklist;
```
```suggestion
  SmallSetVector<User *, 32> Worklist;
```
Change the size to match max number of elements maybe?

nikic committed:
Add test for affected value having side effect

nikic committed:
Add test showing inconsistent behavior with assume on global

nikic committed:
Don't handle globals

Otherwise we get inconsistent behavior based on uses in other
functions.

This means we are also guaranteed that all users are Instructions.

nikic committed:
Style tweaks

artagnon commented on llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp:
```diff
@@ -17,13 +18,48 @@ using namespace llvm;
+      return false;
+  }
+
+  for (unsigned Idx = 0; Idx < Worklist.size(); ++Idx) {
+    Instruction *I = Worklist[Idx];
```
```suggestion
  while (!Worklist.empty()) {
    Instruction *I = Worklist.pop_back_val();
```

nikic replied:
This doesn't work because this is a combined worklist + visited set (see the other comment).

artagnon commented on llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp:
```diff
@@ -17,13 +18,48 @@ using namespace llvm;
+  auto AddUsers = [&](Value *V) {
+    for (User *U : V->users()) {
+      // Bail out if we need to inspect too many users.
+      if (Worklist.size() >= 32)
```
```suggestion
      if (Worklist.size() > 32)
```

artagnon commented on llvm/test/Transforms/DropUnnecessaryAssumes/basic.ll:
```diff
@@ -1,6 +1,9 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
 ; RUN: opt -S -passes=drop-unnecessary-assumes < %s | FileCheck %s
 
+declare void @use(i32 %x)
```
Could use llvm.fake.use?

nikic replied:
This is an intrinsic related to debug info preservation, I don't think we should use it more generally.

artagnon reviewed (approved):
After some more thought, I think a collectUsersRecursively is probably not the best idea. LGTM, thanks!
